### DISCUSSION TITLE
Hdf5.WriteCompounds and attributes

### DISCUSSION THREAD

**Discussion by flappah** [LiorBanai/HDF5-CSharp#246] [Q&A] [answered]
I see the method WriteCompounds accepts a dictionary of attributes. I was hoping this would give me the option to add an attribute to the array that is being written. However nothing is being done with the attribute I supply. When I check the source I see there's no handling of the attributes supplied. Am I correct in this? And if so, any chance this feature gets added?

**Comment by LiorBanai** [comment, marked as answer]
Hi @flappah 
do you want to add attribute to existing item or during creation of the item?
Can you add sample code?
It can be added relative simple (if this is not a bug which may be).

**Reply by flappah** [reply]
I'd like to have the attribute added to the dataset that is created at that moment. In this case you can see in the sample I've added that I'm creating a group of waterlevel feature descriptions. The IHO S104 standard dictates it needs an attribute at that specific level with the name 'chunking'. 

For me this code sample is getting to know (a) the standards itself and (b) your HDF5 library. To get a feel for it. This is the line of code where the attributes are needed:

 ```
Dictionary<string, List<string>> attributes = new Dictionary<string, List<string>>();
attributes.Add("chunking", new List<string>() { "0,0" });
Hdf5.WriteCompounds(groupFId, "WaterLevel", waterLevelFeatures, attributes);
```
[HDF5TestCreator.cs.zip](https://github.com/LiorBanai/HDF5-CSharp/files/9964522/HDF5TestCreator.cs.zip)

**Reply by LiorBanai** [reply]
no problem. I'll check and will add the missing code.

**Reply by flappah** [reply]
Thanks Lior. Really appreciate it!

**Reply by LiorBanai** [reply]
Ok I see the attribute code is missing. working on it.

**Reply by LiorBanai** [reply]
@flappah I have created and issue for this discussion #248

**Comment by LiorBanai** [comment]
Hi @flappah 
is this output what you need (using the above code)? If so I'll publish new version..
[TestAttributesCreation.zip](https://github.com/LiorBanai/HDF5-CSharp/files/9990213/TestAttributesCreation.zip)

**Reply by flappah** [reply]
Hi LiorBanai!! Yep, that's exactly what I'm looking for. Looks great! Thanks!!!

**Comment by LiorBanai** [comment]
glad to know. I'll release new nuget tomorrow.

**Reply by LiorBanai** [reply]
I have released new nuget version (1.16.1). Please give it a try and update back if there are issues :)

**Reply by flappah** [reply]
Thanks! I will!

**Reply by flappah** [reply]
[edit]

I'm beginning to suspect there's a minor issue. The attributes are applied to the group-id that is supplied in the call instead of on the dataset that is created with the command. 

In my situation the group is:

`long groupId = Hdf5.CreateOrOpenGroup(fileId, "Group_F");`

and then there's a new dataset created with:

`Hdf5.WriteCompounds(groupFid, "WaterLevel", waterlevelFeatures, attributes);`

The attributes are applied to the groupId instead of to on the WaterLevel-feature dataset. Is it possible to have the attributes applied to the dataset instead of to the group?

**Reply by LiorBanai** [reply]
oh I see. No problem. I'll update it later this week

**Reply by flappah** [reply]
Thanks man! :) really appreciate it!!

**Reply by LiorBanai** [reply]
@flappah 
can you verify that this zip file output has the correct structure?
[TestAttributesCreation.zip](https://github.com/LiorBanai/HDF5-CSharp/files/10039165/TestAttributesCreation.zip)

**Reply by flappah** [reply]
Hi LiorBanai. I see the chunking attribute is applied to the Group_F group. It should be at the WaterLevel dataset if possible. That's what IHO's S104 standard states. For clarity's sake I've added a sample Hdf5 file NOAA has distributed. 
[104US00_ches_dcf2_20190606T12Z.h5.zip](https://github.com/LiorBanai/HDF5-CSharp/files/10040448/104US00_ches_dcf2_20190606T12Z.h5.zip)

**Reply by LiorBanai** [reply]
Hhh third try.. Will update 🙂

**Reply by flappah** [reply]
> Hhh third try.. Will update 🙂

heheh .. no problem. I did check the source code earlier and I saw that the dataset was closed in the method. For that reason I couldn't add the attributes myself afterwards.

**Reply by LiorBanai** [reply]
Missed a place in the code.
![image](https://user-images.githubusercontent.com/36262995/202702473-33804dfc-0157-4b09-a901-0ee50987d95f.png)
now should be ok.
[TestAttributesCreation3.zip](https://github.com/LiorBanai/HDF5-CSharp/files/10041135/TestAttributesCreation3.zip)

**Reply by flappah** [reply]
Yep, I can confirm it's exactly how I would like to have it. Thanks! Looks perfect!

**Reply by LiorBanai** [reply]
Great. I'll release new version later today (need to go for some errands)

**Reply by flappah** [reply]
> Great. I'll release new version later today (need to go for some errands)

Thanks man! No worries. I'm in no particular hurry since we're currently only testing. I really appreciate your swift efforts!

**Reply by LiorBanai** [reply]
I have release new version (1.16.2) . Give it a try.

**Reply by flappah** [reply]
> I have release new version (1.16.2) . Give it a try.

Thanks LiorBanai! I will. Thanks for your support!

**Reply by flappah** [reply]
I can confirm it works like a charm!! Thanks man!
